lib/bibliothecary/parsers/pypi.rb

```diff
@@ -90,9 +93,28 @@ def self.parse_pipfile(file_contents, options: {})
         map_dependencies(manifest['packages'], 'runtime') + map_dependencies(manifest['dev-packages'], 'develop')
       end
 
+      def self.parse_pyproject(file_contents, options: {})
+        deps = []  
+
+        file_contents = Tomlrb.parse(file_contents)
+        
+        # Parse poetry [tool.poetry] deps
+        poetry_manifest = file_contents.fetch('tool', {}).fetch('poetry', {})
+        deps += map_dependencies(poetry_manifest['dependencies'], 'runtime')
+        deps += map_dependencies(poetry_manifest['dev-dependencies'], 'develop')
+
+        # Parse PEP621 [project] deps
+        pep621_manifest = file_contents.fetch('project', {})
+        pep621_deps = pep621_manifest.fetch('dependencies', []).map { |d| parse_pep_508_dep_spec(d) }
+        deps += map_dependencies(pep621_deps, 'runtime')
```

Comment:
should these be mutually excluded or is it possible for a single manifest to specify both of these dependencies?

Reply:
PEP621 is the "tool-agnostic" way to declare build system requirements in pyproject.toml, and from what I understand Poetry is the only build system that doesn't support it yet. They're discussing it tho, and it looks like one of the examples includes both the "tool.poetry" And "project" namespaces for declaring deps: https://github.com/python-poetry/roadmap/issues/3

Reply:
ah yeah, just found an example here. The deps are the same in both section, but still seems safer to combine them both until we have a reason not to: https://sourcegraph.com/github.com/IceTiki/ruoli-sign-optimization/-/blob/pyproject.toml?subtree=true

Reply:
In that case, does dep need to be deduped? I'm not sure what clients of this method are expecting

Reply:
ah yes! Good call, I think we do that higher up in `Bibliothecary.analyse_file` but it does seem like some of the platform-specific parsers will uniq too.